### PR TITLE
S3168: Broaden detection of event handlers

### LINKED ISSUE #1181: Rule S3168: Broaden detection of event handlers
### Description

There are several frameworks which do not follow the standard signature for event handlers of what are conceptually events.

Here are 2 examples off the top of my head:
1. The Reactive Extensions' IObservable.Subscribe extension method that takes a delegate
2. Prism's EventAggregator and the region navigation callbacks (OnNavigatedTo, etc)

I believe that the approach of the current implementation of S3168 is far too restrictive and leads very easily to disabling this rule altogether.

May I suggest considering methods whose name starts with 'On' as event handlers, and therefore acceptable for async void?
(Of course it would have to be a bit less naïve, like matching with something along the lines of `On[^a-z]\w*` to avoid accepting `OnboardClient()`) 
That would force the user to reconsider using async void if such a prefix were not suitable for the method.

### Repro steps

1. Create a method `OnPersonRegistered` that handles an event raised by Prism's EventAggregator.
2. Make `OnPersonRegistered` await an asynchronous method (`Task.Yield()`, for instance) so that it requires a signature with async void.
3. Run analysis with rule S3168 activated

### Expected behavior

The method `OnPersonRegistered` being the handler of what is conceptually and by name (**On**Something) an event, the async void in its signature is unavoidable and should be acceptable.
Therefore, rule S3168 should not be considered violated.

### Actual behavior

The analyzer raises a violation for rule S3168.

### Related information

* SonarC# Version 6.8.1.4648
* Visual Studio Version 15.5.4

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #1181

### CODE REVIEW

**Comment by pavel-mikula-sonarsource** [comment]
Let's merge this one even with failed QG. It's coming from cleanup of old code.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/AsyncVoidMethod.cs [hunk 1]
```diff
@@ -48,7 +48,7 @@ private async void EventHandlerCases_MyEvent(object sender, bool e)
             // Property changed implementation goes here
         }
 
-        private async void ArbreDesClefs_ItemInvoked(TreeView sender, TreeViewItemInvokedEventArgs args) { }    // Noncompliant FP
+        private async void ArbreDesClefs_ItemInvoked(TreeView sender, TreeViewItemInvokedEventArgs args) { }    // Compliant, doesn't inherit from EventArgs, but looks like it
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/AsyncVoidMethod.cs:60]: what if ` private async void ArbreDesClefs_ItemInvoked(TreeView somethingElse, TreeViewItemInvokedEventArgs arguments)` ? is using `sender` and `args` a pattern enforced by MS guidelines?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/AsyncVoidMethod.cs [hunk 2]
```diff
@@ -33,6 +34,27 @@ private async void EventHandlerCases_MyEvent(object sender, bool e)
...
+        {
+            // Property changed implementation goes here
+        }
+
+        private async void ArbreDesClefs_ItemInvoked(TreeView sender, TreeViewItemInvokedEventArgs args) { }    // Noncompliant FP
+
+
+        // Substitute for reference to Xamarin.Forms, Windows.UI.Xaml.Controls
+        public class BindableObject { }
+        public class TreeView { }
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/AsyncVoidMethod.cs:65]: maybe add also the namespace `Windows.UI.Xaml.Controls` and test with more than TreeView or `TreeViewItemInvokedEventArgs` , and rely also if there are items in that namespace?

although it may be better to keep it simple with `EventArgs` suffix

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownMethods.cs [hunk 3]
```diff
@@ -31,194 +31,162 @@ public static bool IsMainMethod(this IMethodSymbol methodSymbol)
...
-                methodSymbol.Parameters.Length == 0 &&
-                methodSymbol.ReturnType.Is(KnownType.System_String);
-        }
+            return methodSymbol != null
+                && !methodSymbol.IsOverride
+                && methodSymbol.IsStatic
+                && methodSymbol.MethodKind == MethodKind.Ordinary
+                && methodSymbol.Name == nameof(object.Equals)
+                && methodSymbol.IsInType(KnownType.System_Object)
+                && HasCorrectArguments()
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownMethods.cs:69]:
```suggestion
                && HasCorrectParameters()
```

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownMethods.cs [hunk 4]
```diff
@@ -31,194 +31,162 @@ public static bool IsMainMethod(this IMethodSymbol methodSymbol)
...
-                methodSymbol.Parameters[1].Type.Is(KnownType.System_Runtime_Serialization_StreamingContext) &&
-                methodSymbol.ReturnsVoid &&
-                (methodSymbol.Name == "GetObjectData" ||
-                methodSymbol.Name == explicitName);
+            return methodSymbol != null
+                && methodSymbol.Parameters.Length == 2
+                && methodSymbol.Parameters[0].Type.Is(KnownType.System_Runtime_Serialization_SerializationInfo)
+                && methodSymbol.Parameters[1].Type.Is(KnownType.System_Runtime_Serialization_StreamingContext)
+                && methodSymbol.ReturnsVoid
+                && (methodSymbol.Name == "GetObjectData" || methodSymbol.Name == explicitName);
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownMethods.cs:120]: I think you can move name checks before the Parameters Type checks, right? 

and maybe the same for all name checks which are done after `Type.Is()`  (e.g. lines 109,178,182) - string comparisons should be cheaper than the type checks which eventually [do multiple comparisons](https://github.com/SonarSource/sonar-dotnet/blob/6ef1c842c09232c065b477f79910effb8554523f/analyzers/src/SonarAnalyzer.Common/Helpers/TypeHelper.cs#L60)

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/AsyncVoidMethod.cs [hunk 5]
```diff
@@ -35,15 +35,24 @@ private async void EventHandlerCases_MyEvent(object sender, bool e)
...
+        static async void On() { }      // Noncompliant
+        static async void Onboard() { } // Noncompliant
+        static async void ToX() { }     // Noncompliant
+        static async void OnX() { }
+
+        static async void OnŘád() { }
+        static async void Onřád() { }   // Noncompliant
+
+        static async void OnΘ() { }     // Compliant, Uppercase Theta
+        static async void Onθ() { }     // Noncompliant, Lowercase Theta
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/AsyncVoidMethod.cs:52]: also 

- `static async void On3People` 
- `static async void ONCAPS` 
- `static async void On_Underscore`
- `static async void onEvent`

File: analyzers/rspec/cs/S3168_c#.html [hunk 6]
```diff
@@ -47,6 +47,7 @@ <h2>Compliant Solution</h2>
 }
 </pre>
 <h2>Exceptions</h2>
-<p>Event handlers, i.e. methods with two arguments, first one matching <code>object sender</code> and the second being or inheriting from
-<code>EventArgs</code>, are ignored.</p>
+<p>Event handlers, i.e. methods with two arguments, first one matching type <code>object</code> or name <code>sender</code> and the second being or
+inheriting from <code>EventArgs</code>, are ignored.</p>
+<p>Event handlers named as <code>OnSomething</code> are also ignored.</p>
```
Comment (andrei-epure-sonarsource) [inline analyzers/rspec/cs/S3168_c#.html:52]: I would say "Methods named as `OnSomething` ...", right? We don't know for sure they're event handlers?

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/AsyncVoidMethod.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/AsyncVoidMethod.cs:60]: Can you please clarify this question in more detail? I didn't understand it.
I'll update the comment a little, as the actual fix was about the sender args.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/AsyncVoidMethod.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/AsyncVoidMethod.cs:65]: There [A LOT](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls?view=winrt-19041) of types ending with `EventArgs`, so for Xamarin, it should be fine with the suffix. 

I wanted to stay Xamarin-reference-free to have an universal solution. With any class as a sender and any `WhateverNameEndingWithEventArgs` argument (as supported by the old implementation)

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownMethods.cs [hunk 7]
```diff
@@ -66,58 +66,58 @@ public static bool IsStaticObjectEquals(this IMethodSymbol methodSymbol)
...
             && methodSymbol.Name == nameof(object.ToString)
+            && (methodSymbol.IsOverride || methodSymbol.IsInType(KnownType.System_Object))
             && methodSymbol.Parameters.Length == 0
             && methodSymbol.ReturnType.Is(KnownType.System_String);
 
         public static bool IsIDisposableDispose(this IMethodSymbol methodSymbol)
         {
             const string explicitName = "System.IDisposable.Dispose";
             return methodSymbol != null
+                && (methodSymbol.Name == nameof(IDisposable.Dispose) || methodSymbol.Name == explicitName)
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownMethods.cs:98]: I'm not sure this change makes sense - checking a bool or int property should be less expensive than checking the name

(I guess you can make this change in a different PR if you want)

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownMethods.cs (hunk 7 above)
Comment (andrei-epure-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/KnownMethods.cs:98]: (discussed offline - the changes were made to keep a common checking pattern  - first name, then params)
